smartcontract/service/statemachine.go

```diff
@@ -0,0 +1,287 @@
+package service
+
+import (
+	"DNA/common"
+	"DNA/core/asset"
+	"DNA/core/code"
+	"DNA/core/contract"
+	"DNA/core/ledger"
+	"DNA/core/store"
+	"DNA/core/transaction"
+	"DNA/crypto"
+	"DNA/errors"
+	. "DNA/smartcontract/errors"
+	"DNA/smartcontract/states"
+	"DNA/smartcontract/storage"
+	"DNA/vm/avm"
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math"
+)
+
+type StateMachine struct {
+	*StateReader
+	CloneCache *storage.CloneCache
+}
+
+func NewStateMachine(dbCache storage.DBCache, innerCache storage.DBCache) *StateMachine {
+	var stateMachine StateMachine
+	stateMachine.CloneCache = storage.NewCloneDBCache(innerCache, dbCache)
+	stateMachine.StateReader = NewStateReader()
+	stateMachine.StateReader.Register("Neo.Validator.Register", stateMachine.RegisterValidator)
+	stateMachine.StateReader.Register("Neo.Asset.Create", stateMachine.CreateAsset)
+	stateMachine.StateReader.Register("Neo.Contract.Create", stateMachine.CreateContract)
+	stateMachine.StateReader.Register("Neo.Blockchain.GetContract", stateMachine.GetContract)
+	stateMachine.StateReader.Register("Neo.Asset.Renew", stateMachine.AssetRenew)
+	stateMachine.StateReader.Register("Neo.Storage.Get", stateMachine.StorageGet)
+	stateMachine.StateReader.Register("Neo.Contract.Destroy", stateMachine.ContractDestory)
+	stateMachine.StateReader.Register("Neo.Storage.Put", stateMachine.StoragePut)
+	stateMachine.StateReader.Register("Neo.Storage.Delete", stateMachine.StorageDelete)
+	stateMachine.StateReader.Register("Neo.Contract.GetStorageContext", stateMachine.GetStorageContext)
+	return &stateMachine
+}
+
+func (s *StateMachine) RegisterValidator(engine *avm.ExecutionEngine) (bool, error) {
+	pubkeyByte := avm.PopByteArray(engine)
+	pubkey, err := crypto.DecodePoint(pubkeyByte)
+	if err != nil {
+		return false, err
+	}
+	if result, err := s.StateReader.CheckWitnessPublicKey(engine, pubkey); !result {
+		return result, err
+	}
+	b := new(bytes.Buffer)
+	pubkey.Serialize(b)
+	validatorState, err := s.CloneCache.GetInnerCache().GetOrAdd(store.ST_Validator, b.String(), &states.ValidatorState{PublicKey: pubkey})
```

Comment:
this execution path may have nil pointer dereference problem:  GetOrAdd(store.ST_Validator... ---> DBCache.TryGetInternal -> states.GetStateValue --> StatesMap[ST_Validator]

Reply:
yes, StatesMap forget store.ST_Validator mapping